📖 Add documentation for declarative plugin (#2562)
kubernetes-sigs/kubebuilder: docs/book/src/plugins/declarative-v1.md

> # Declarative Plugin

Reviewer: We need also to add a new section in:

<img width="287" alt="Screenshot 2022-03-26 at 09 43 35" src="https://user-images.githubusercontent.com/7708031/160233990-60d4d4b9-ccf9-46bf-b9ec-2c3966465094.png">

Maybe 
`7. Available Plugins`

With this subsection:

7.1  Declarative Plugin

WDYT?

Reviewer: See: https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/SUMMARY.md

Reviewer: @mariuskimmina, 

> also, should it be 7. Available Plugins or 6.3 Available Plugins? I think I like to keep them in the general Plugins section

Yes, that makes sense 👍

Author: I have added an available plugins section now

The declarative plugin is for building declarative cluster operators with kubebuilder.
Declarative operators provide a fast path to orchestrating Kubernetes deployments to enable domain experts to focus
on their component instead of re-answering questions like 'How do I get this YAML into the cluster?' or 'How do I
update it?'.  
With the declarative plugin manifests are created under `channels/packages/<packagename>/<version>/manifest.yaml` 
and they will be loaded and applied to the cluster. An example of the pattern can be found [here](https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata/project-v3-addon)
The `channels/stable` file declares the current stable version that will be used by default.

## Advantages

* Working with manifests lets us release a new app version without needing a new operator version
* The declarative manifest makes it easier for users to understand what is changing in each version
* It should result in less / simpler code

> 

Reviewer: WDYT, we added something like: 

## How to use:

### If you want that any API(s) and its respective controller(s) generate to reconcile them of your project adopt this partner then:
Use these plugin when you create a new project from the scratch, e.g:

```
kb init --plugins=declarative/v1
```

### If you want to adopt this partner for specific API(s) and its respective controller(s) generate to reconcile then: 
Use this plugin when you are creating the api

kb init create api  "todo: add e.g options from quick start" --plugins=declarative/v1

## Subcommands

> The declarative plugin implements the following subcommands:

Reviewer: ```suggestion
The declarative plugin implements the following subCommands:
```

* init
* create api

## Affected files

> The following scaffolds will be affected by the plugin:

Reviewer: ```suggestion
The following scaffolds will be created or updated by this plugin:
```

* `controllers/*_controller.go`
* `api/*_types.go`
* `channels/packages/<packagename>/<version>/manifest.yaml`
* `channels/stable`

> 

Reviewer: ```suggestion
* Dockerfile
```
See : https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v3-addon/Dockerfile#L16-L19

# Declarative Plugin

> The declarative plugin is for building declarative cluster operators with kubebuilder.

Reviewer: ```suggestion
The declarative plugin allows you create [controllers][controller-runtime] using the project [kubebuilder-declarative-pattern][kubebuilder-declarative-pattern]. By using these plugins Kubebuilder users are able to scaffold the projects files following its partner.  
```

> Declarative operators provide a fast path to orchestrating Kubernetes deployments to enable domain experts to focus
> on their component instead of re-answering questions like 'How do I get this YAML into the cluster?' or 'How do I
> update it?'.  
> With the declarative plugin manifests are created under `channels/packages/<packagename>/<version>/manifest.yaml` 
> and they will be loaded and applied to the cluster. An example of the pattern can be found [here](https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata/project-v3-addon)
> The `channels/stable` file declares the current stable version that will be used by default.

Reviewer: ```suggestion
## When should I use this plugin

- If you are looking to scaffold one or more [controller's][controller-runtime] following its partner ( See an e.g. of the reconcile method implemented with. [here][addon-v3-controller]) 
- Also, if you are looking to have the scaffolded files shipped inside of your Manager container then, it might a good option for your use case. See on the Dockerfile the declarative plugin works with channels to allow you to push the manifests. [More info][dockerfile-addon] 
```

Reviewer: ```
[dockerfile-addon]: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v3-addon/Dockerfile#L16-L19
```

Author: I think I would like to keep line 4 to 6 here as a general introduction / overview of the plugins purpose - I would remove line 7 to 9 and add your suggested 'When to use it' section, would you be okay with that?

Author: `[here][addon-v3-controller]` is not defined - where do you want to link to here?

Author: > I think I would like to keep line 4 to 6 here as a general introduction / overview of the plugins purpose - I would remove line 7 to 9 and add your suggested 'When to use it' section, would you be okay with that?

Okay, never mind that, I just saw you suggested changes to the introduction as well - ignore the above comment

## How to use it

> The declarative plugin can only be used in combination with another plugin that scaffolds the API and controller - such as the `go/v3` plugin.

Reviewer: ```suggestion
The declarative plugin requires to be used with one of the available Golang plugins. 
```

> If you want that any API(s) and its respective controller(s) generate to reconcile them of your project adopt this partner then:  

Reviewer: ```suggestion
- If you want that any API(s) and its respective controller(s) generate to reconcile them to your project adopt this partner then:  
```

> Use these plugin when you create a new project from the scratch, e.g:

Reviewer: ```suggestion
```

> ```

Reviewer: ```suggestion
```sh
```

kubebuilder init --plugins=go/v3,declarative/v1 --domain example.org --repo example.org/guestbook-operator
```

> If you want to adopt this pattern for specific API(s) and its respective controller(s) then:  

Reviewer: ```suggestion
- If you want to adopt this pattern for specific API(s) and its respective controller(s) (not for any API/controller scaffold using Kubebuilder CLI) then:  
```

> Use this plugin when you are creating the API

Reviewer: ```suggestion
```

> ```

Reviewer: ```suggestion
```sh
```

kubebuilder create api --plugins=go/v3,declarative/v1 --version v1 --kind Guestbook
```

> ## Advantages
> 
> * Working with manifests lets us release a new app version without needing a new operator version
> * The declarative manifest makes it easier for users to understand what is changing in each version
> * It should result in less / simpler code

Reviewer: ```suggestion
```
I think would be hard for us to have this section as a standard.
Maybe when should I use it shows to fit better wdyt?

## Subcommands

The declarative plugin implements the following subcommands:

> * init

Reviewer: ```suggestion
* init ($ kubebuilder init [OPTIONS])
```

> * create api

Reviewer: ```suggestion
* create api ($ kubebuilder create api [OPTIONS])
```

## Affected files

The following scaffolds will be created or updated by this plugin:

* `controllers/*_controller.go`
* `api/*_types.go`
* `channels/packages/<packagename>/<version>/manifest.yaml`
* `channels/stable`
* `Dockerfile`

## Further ressources

* Read more about the [declarative pattern](https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern)
* Watch the KubeCon 2018 Video [Managing Addons with Operators](https://www.youtube.com/watch?v=LPejvfBR5_w)

> * Check the [implementation](https://github.com/kubernetes-sigs/kubebuilder/tree/master/pkg/plugins/golang/declarative)

Reviewer: ```suggestion
* Check the plugin [implementation](https://github.com/kubernetes-sigs/kubebuilder/tree/master/pkg/plugins/golang/declarative)

[dockerfile-addon]: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v3-addon/Dockerfile#L16-L19
[addon-channels-info]: https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/master/docs/addon/walkthrough/README.md#adding-a-manifest
[controller-runtime]: https://github.com/kubernetes-sigs/controller-runtime
[kubebuilder-declarative-pattern]: https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern
[testdata]: https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata/
```

# Declarative Plugin

The declarative plugin allows you to create [controllers][controller-runtime] using the [kubebuilder-declarative-pattern][kubebuilder-declarative-pattern]. 
By using the declarative plugin, you can make the required changes on top of what is scaffolded by default when you create a Go project with Kubebuilder and the Golang plugins (i.e. go/v2, go/v3). 

> <aside class="note">
> You can check samples using this plugin by looking at the `project-<go plugin version>-addon` samples under the [testdata][testdata] directory on the root directory of the Kubebuilder project.
> </aside> 

Reviewer: ```suggestion
<aside class="note">
<h1>Samples</h1>

You can check samples using this plugin by looking at the `project-<go plugin version>-addon` samples under the [testdata](https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata/) directory on the root directory of the Kubebuilder project.

</aside>
```

Seems that the note does not work with the link alias :-(

Author: That is strange, because it works in `plugins.md` - there an alias is also used inside the <aside> tag
